pkg/experiment/metastore/index/index.go

```diff
@@ -66,12 +68,28 @@ type Store interface {
 
 type Index struct {
 	logger log.Logger
-	config *Config
+	config Config
 	store  Store
 
-	mu               sync.Mutex
-	loadedPartitions map[tenantPartitionKey]*indexPartition
+	// The global lock protects index partitions.
+	// Index partition shards have their own locks.
+	global sync.Mutex
+	// TODO: cache at the shard level.
+	tenantPartitions map[tenantPartitionKey]*indexPartition
 	partitions       []*store.Partition
```

Comment:
I'm going to redesign the index soon, so this PR does not change its structure; the only goal is to avoid long locks. Ideally, we should separate the mutable write buffer – fully in-memory – from the read buffer – an immutable on-disk/in-store part that is loaded on demand. Queries should ideally be delegated to query-backend workers.